src/input/HidApiInputDriver.cc

```diff
@@ -58,9 +67,55 @@ static const struct device_id device_ids[] = {
     { 0x256f, 0xc631, &HidApiInputDriver::hidapi_decode_axis2, &HidApiInputDriver::hidapi_decode_button2, "3Dconnexion Space Mouse Pro Wireless (cabled)"},
     { 0x256f, 0xc632, &HidApiInputDriver::hidapi_decode_axis2, &HidApiInputDriver::hidapi_decode_button2, "3Dconnexion Space Mouse Pro Wireless"},
     { 0x256f, 0xc635, &HidApiInputDriver::hidapi_decode_axis2, &HidApiInputDriver::hidapi_decode_button2, "3Dconnexion Space Mouse Compact"},
-    { -1, -1, NULL, NULL, NULL},
+    // This is reported to be used with a 3Dconnexion Space Mouse Wireless 256f:c62e
+    { 0x256f, 0xc652, &HidApiInputDriver::hidapi_decode_axis2, &HidApiInputDriver::hidapi_decode_button2, "3Dconnexion Universal Receiver"},
+    { -1, -1, nullptr, nullptr, nullptr},
 };
 
+#define HIDAPI_LOG(f) hidapi_log(boost::format(f))
+#define HIDAPI_LOGP(f,a) hidapi_log(boost::format(f) % a)
+
+static void hidapi_log(boost::format format) {
+	if (logstream) {
+		logstream << format.str() << std::endl;
+		if (logstream.tellp() > MAX_LOG_SIZE) {
+			logstream.close();
+		}
+	}
+}
+
+static void hidapi_log_input(unsigned char *buf, int len)
+{
+	if (logstream) {
+		std::stringstream s;
```

Comment:
Minor detail: `std::ostringstream` is usually preferred when you never read from it.

Reply:
done